### ISSUE TITLE
Two language support?

### ISSUE THREAD

**Issue by fantasticfears** [adamvoss/vscode-languagetool#9]
I write English and Chinese but I can only set the language to one. Is it possible to have an array for spell checking?

**Comment by adamvoss** [comment]
How would the software know which to use for each passage?

If both languages are checked, I imagine English passages would have a bunch of errors from the Chinese check-er and vice-versa.

**Comment by fantasticfears** [comment]
In a situation like Chinese and English, Unicode Multilingual Plane can help to determine what language script the text it is. For EU languages, it may be harder.

> If both languages are checked, I imagine English passages would have a bunch of errors from the Chinese check-er and vice-versa.

For example, I use English in my OneNote which has the same configuration setting like LT, so it gives me a screen of errors everytime since a lot of text is in Chinese. To my view, I could imagine using more languages is quite common in today's world.

**Comment by adamvoss** [comment]
I understand the concern, but I would need to be presented a technical solution that would be generally workable before I would try implementing this myself.  Though, I will consider any PRs on their merits.

> Unicode Multilingual Plane can help to determine what language script the text it is.

As you mentioned this _could_ work to help with English vs Chinese, but it would not apply generally.  As I understand it, Unicode made a mess of CJK (Chinese/Japanese/Korean) scripts so even if we identified one of them we would still need to fall back on your array of enabled language and hope there is only one applicable.

It sounds like you want automatic language detection.  I could see this being potentially problematic in the context of a grammar checker since poorly-formed text may not correctly identify with the intended language.  Putting that concern aside, a language identification library would still need to be identified and rules proposed with respect to how to split the text for language identification.

**Comment by fantasticfears** [comment]
> I understand the concern, but I would need to be presented a technical solution that would be generally workable before I would try implementing this myself. Though, I will consider any PRs on their merits. 

Of course, it's worth for discussion before heading into implementation.
Considered the fact, vscode is often used for programmers. So who uses languagetool are probably programmers or related with technical field. I wouldn't believe many will write CJK and English as the same time.
This reason of course is not strong enough, but it sounds practical approach to me.

But if language server is not subject to do so, if it has some context in analyzing sentences, Unicode isn't enough. But I don't think it's possible to do so called automatic language detection with a little cost. I would refrain to build such library in the first place. On internet, encoding is a lead to language but it's not perfect. Either detection is hard or UTF-8 is used. In HTTP, relying on accept-language header is better. But text editor only relies on the user setting.

In conclusion, I didn't find a way to detect languages at a small cost. But a word level spell checker for combining language is still practical for some users if language server does that.

**Comment by adamvoss** [comment]
In general, I understand what you have written; however, I'll admit I have a little bit of trouble understanding some of the specifics and I do not want to miss something.

To be sure I understand, the obvious places to "split" the language choice are:

- Global (via the default or via user settings, currently exists)
- Project/Workspace (via workspace settings, currently exists)
- File
- Paragraph
- Sentence
- Word

Which level(s) do you think are needed?

___
> But a word level spell checker for combining language is still practical for some users if language server does that.

That is an interesting idea, but also sounds like a different project.  My intentions in bringing LanguageTool to VSCode was to have ___grammar___ checking.  I honestly considered disabling the spelling rules because there are other spell checkers for VSCode, but no one has complained about them yet.

Since LanguageTool does have spell checking capabilities, I suspect you could use it to accomplish your multi-language spell checker.  You could even fork this project to do so.  You would want to disable all the _grammar_ rules, check the text against every enabled language, then inspect the results on a per word basis.  If one says valid valid mark it as valid, otherwise mark the spelling error. (Note: this assumes both language spell checkers present unknown words as errors.)

**Comment by fantasticfears** [comment]
Thanks for your patience, I usually go on too fast in a conversion. Your questions and clarification make sense to me. I think I misunderstood your first intention.

To my particular need, I'm settled with the workspace level setting for my things at hand. Sometimes, file level settings come in handy. For example, a Jekyll site with multiple languages in different languages.

As of spell checker, it's quite common to me that combining English and Chinese. For example, when I write a technical post in Chinese, it's inevitable to have abbreviations in a Chinese sentences.  I wouldn't expect the grammar checker works in this case. But well, spell checkers, they are really useful.

> Since LanguageTool does have spell checking capabilities, I suspect you could use it to accomplish your multi-language spell checker. You could even fork this project to do so. You would want to disable all the grammar rules, check the text against every enabled language, then inspect the results on a per word basis. If one says valid valid mark it as valid, otherwise mark the spelling error. (Note: this assumes both language spell checkers present unknown words as errors.)

I have no doubt it will give me too many false positives...I don't even notice the difference since I have installed your plugin yesterday. You have every right to make the decision towards disabling the spelling rules feature you've mention. Anyway it works so COOL, thank you.
